### DISCUSSION TITLE
cargo doc --open of pyo3

### DISCUSSION THREAD

**Discussion by wangjiawen2013** [PyO3/pyo3#4354] [Questions] [answered]
Hi,
I wrote a python module using pyo3, but I find that I cannot generate the documentation using rust doc comments.
Cargo.toml:
```
[package]
name = "string-sum"
version = "0.1.0"
edition = "2018"

[lib]
name = "string_sum"
# "cdylib" is necessary to produce a shared library for Python to import from.
#
# Downstream Rust code (including code in `bin/`, `examples/`, and `tests/`) will not be able
# to `use string_sum;` unless the "rlib" or "lib" crate type is also included, e.g.:
# crate-type = ["cdylib", "rlib"]
crate-type = ["cdylib"]

[dependencies.pyo3]
version = "0.15.1"
features = ["extension-module"]
```

src/libs.rs:
```
use pyo3::prelude::*;

/// Formats the sum of two numbers as string.
#[pyfunction]
fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
    Ok((a + b).to_string())
}

/// A Python module implemented in Rust. The name of this function must match
/// the `lib.name` setting in the `Cargo.toml`, else Python will not be able to
/// import the module.
#[pymodule]
fn string_sum(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;

    Ok(())
}
```
Then I want to generate documenation using `cargo doc --open`, but what I saw is:
![image](https://github.com/user-attachments/assets/9e995bff-b622-4478-8563-e86dc4ba847c)
As you can see, it's very different from others rust crates documentation.

**Comment by davidhewitt** [comment]
What result are you expecting to get here? The functions you've defined are Python functions, so it seems correct to me that they don't have Rust API documentation.

**Comment by birkenfeld** [comment]
In 0.22 you also don't get the PyInit entry anymore. 

However, when setting the functions as `pub`, you'll see them in rustdoc as expected. 

So, I don't see a problem here.

**Comment by wangjiawen2013** [comment]
I mistakenly thought it will generate rust API documentation. Could I see the comments in the Python function/module ?

**Reply by birkenfeld** [reply, marked as answer]
You're not mistaken in that, but the functions need to be `pub` for that.

**Reply by wangjiawen2013** [reply]
Thank you, I can see it now!
